### PR TITLE
Align pr-agent workflow tests with trigger job key

### PR DESCRIPTION
**Pull request by mohavro**
### **Auto-created Ticket**

[#400](https://github.com/mohavro/financial-asset-relationship-db/issues/400)

## Summary
- update pr-agent workflow integration tests to assert the existing `pr-agent-trigger` job key
- clean up malformed blocks in the pr-agent workflow tests and restore missing typing imports
- fix indentation in the environment variable helper to allow test collection

## Testing
- python -m pytest tests/integration/test_github_workflows.py -k pr_agent_has_trigger_job

[Codex Task](https://chatgpt.com/codex/tasks/task_e_691ee572951c8324bbb5fb1b2480d08f)

## Summary by Sourcery

Update pr-agent workflow integration tests to assert the 'pr-agent-trigger' job key, tidy up test blocks and docstrings, and correct helper indentation for env var validation.

Enhancements:
- Align pr-agent workflow integration tests to use the 'pr-agent-trigger' job key

Tests:
- Clean up malformed docstrings and restore missing typing imports in workflow tests
- Fix indentation in the environment variable helper to enable test collection

---
## Summary by cubic
Aligns pr-agent workflow tests to the 'pr-agent-trigger' job and fixes malformed tests so pytest runs. Adds a workflows–requirements integration suite and hardens checks: standard Ubuntu runner, non-empty checkout token, integer non‑negative fetch‑depth, exactly one setup‑python with python-version 3.11, logical setup order, pinned action versions, valid env var names, well‑formed/indented with: blocks with no orphans, unique step names, no duplicate YAML keys; plus requirements‑dev hygiene (PyYAML and types‑PyYAML present with matching majors, install‑before‑test, Python 3.8+, PEP 440‑valid specifiers, clean formatting incl. final newline/no trailing spaces, no duplicate packages, consistent operators, essential dev tools).

<sup>Written for commit 7358a5ac745f92bcbfb173fe71c29592fdbe471c. Summary will update automatically on new commits.</sup>

<details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt t
...(truncated)

___

## **CodeAnt-AI Description**
Harden workflow and requirements validation tests

### What Changed
- Workflow validation suite now requires PyYAML, skips cleanly when it is absent, and explicitly asserts the pr-agent workflow defines the `pr-agent-trigger` job with proper checkout fetch depth, Python/Node setup, and secrets handling.
- Added integration tests that tie `.github/workflows` to `requirements-dev.txt`, checking that workflows install dependencies before running tests, PyYAML and types-PyYAML can parse the workflows, and essential tooling (pytest, flake8, black, mypy) and compatible Python versions are listed in requirements-dev.
- Helper tests now enforce uppercase alphanumeri
...(truncated)

### CODE REVIEW

File: tests/integration/test_github_workflows.py [hunk 1]
```diff
@@ -1158,12 +1134,12 @@ def check_env_vars(env_dict):
...
-                is_valid_chars = all(c.isalnum() or c == '_' for c in key)
-                if not key.isupper() or not is_valid_chars:
-                    invalid.append(key)
-            return invalid
+        for key in env_dict.keys():
+            # Ensure all characters are either alphanumeric or underscore
+            is_valid_chars = all(c.isalnum() or c == '_' for c in key)
+            if not key.isupper() or not is_valid_chars:
+                invalid.append(key)
+        return invalid
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]: @qodo-merge-pro fix this

File: tests/integration/test_requirements_dev.py [hunk 2]
```diff
@@ -302,4 +302,203 @@ def test_existing_packages_preserved(self, requirements: List[Tuple[str, str]]):
...
+        duplicates = []
+        for pkg in package_names:
+            if pkg in seen:
+                duplicates.append(pkg)
+            seen.add(pkg)
+        
+        assert len(duplicates) == 0, (
+            f"Duplicate packages found in requirements-dev.txt: {duplicates}. "
+            "Each package should appear only once."
+        )
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_dev.py]: @qodo-merge-pro fix this

File: tests/integration/test_requirements_dev.py [hunk 3]
```diff
@@ -302,4 +302,203 @@ def test_existing_packages_preserved(self, requirements: List[Tuple[str, str]]):
...
+            "PyYAML must be in requirements-dev.txt as it's needed for workflow validation tests"
+        )
+        
+        # Verify we can actually import yaml (validates the requirement works)
+        try:
+            import yaml
+            # Successfully imported - requirement is satisfied
+            assert yaml.safe_load("key: value") == {'key': 'value'}, "PyYAML import successful"
+        except ImportError:
+            pytest.skip("PyYAML not installed in test environment (will be installed from requirements)")
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_dev.py]: @qodo-merge-pro fix this

File: tests/integration/test_workflow_requirements_integration.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            for sep in ('>=', '==', '<=', '>', '<', '~='):
+                if sep in line:
+                    pkg = line.split(sep)[0].strip()
+                    version = line.split(sep)[1].strip()
+                    requirements.append((pkg, f"{sep}{version}"))
+                    break
+            else:
+                requirements.append((line.strip(), ''))
+    
+    return requirements
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_workflow_requirements_integration.py]: @qodo-merge-pro fix this

File: tests/integration/test_workflow_requirements_integration.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+            try:
+                req = Requirement(line)
+                # Only include the name and version specifier; ignore extras and markers for matching
+                requirements.append((req.name, str(req.specifier)))
+            except Exception:
+                # Fall back to previous behavior on unparseable lines
+                # Keep the raw (stripped) line with empty spec to avoid false negatives
+                requirements.append((line, ''))
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_workflow_requirements_integration.py]: @qodo-merge-pro fix this

File: tests/integration/test_github_workflows.py [hunk 6]
```diff
@@ -342,24 +344,13 @@ def test_pr_agent_checkout_has_token(self, pr_agent_workflow: Dict[str, Any]):
...
         review_job = pr_agent_workflow["jobs"]["pr-agent-trigger"]
         steps = review_job.get("steps", [])
-        
+
         python_steps = [
-            s for s in steps 
+            s for s in steps
             if s.get("uses", "").startswith("actions/setup-python")
         ]
         assert len(python_steps) > 0, "Review job must set up Python"
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
        review_job = pr_agent_workflow["jobs"]["pr-agent-trigger"]
        steps = review_job.get("steps", [])

        python_steps = [
            s for s in steps
            if s.get("uses", "").startswith("actions/setup-python")
        ]
        assert len(python_steps) > 0, "Review job must set up Python"
```

File: tests/integration/test_workflow_requirements_integration.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                break
+        
+        if python_version:
+            # Verify it's a supported version (3.8+)
+            version_parts = str(python_version).split('.')
+            if len(version_parts) >= 2:
+                major = int(version_parts[0])
+                minor = int(version_parts[1])
+                
+                assert major >= 3 and minor >= 8, (
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_workflow_requirements_integration.py]:
```suggestion
                assert major >= 3 and minor >= 8, (
```

File: tests/integration/test_workflow_requirements_integration.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+class TestPyYAMLAvailability:
+    """Test that PyYAML is properly configured for use in workflow tests."""
+    
+    def test_pyyaml_in_requirements_for_workflow_validation(self):
+        """Test that PyYAML is in requirements-dev.txt for workflow validation tests."""
+        assert REQUIREMENTS_FILE.exists(), "requirements-dev.txt not found"
+        
+        requirements = parse_requirements(REQUIREMENTS_FILE)
+        package_names = [pkg for pkg, _ in requirements]
+
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_workflow_requirements_integration.py]:
```suggestion
        package_names = [pkg for pkg, _ in requirements]
        
```

File: tests/integration/test_workflow_requirements_integration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_workflow_requirements_integration.py]:
```suggestion
def parse_requirements(file_path: Path) -> List[Tuple[str, str]]:
    """Parse requirements file and return list of (package, version_spec) tuples."""
    requirements = []
    
    with open(file_path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            
            if not line or line.startswith('#'):
                continue
            
            # Extract package name before any version specifier
            for sep in ('>=', '==', '<=', '>', '<', '~='):
                if sep in line:
                    pkg = line.split(sep)[0].strip()
                    version = line.split(sep)[1].strip()
                    requirements.append((pkg, f"{sep}{version}"))
                    break
            else:
                requirements.append((line.strip(), ''))
    
    return requirements
```

File: tests/integration/test_requirements_dev.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_dev.py]:
```suggestion
    def test_no_duplicate_package_names(self):
        """Test that no package appears multiple times in requirements-dev.txt."""
        assert REQUIREMENTS_FILE.exists(), "requirements-dev.txt not found"
        
        requirements = parse_requirements(REQUIREMENTS_FILE)
        
        # Extract package names (case-insensitive)
        package_names = [pkg.lower() for pkg, _ in requirements]
        
        # Find duplicates
        seen = set()
        duplicates = []
        for pkg in package_names:
            if pkg in seen:
                duplicates.append(pkg)
            seen.add(pkg)
        
        assert len(duplicates) == 0, (
            f"Duplicate packages found in requirements-dev.txt: {duplicates}. "
            "Each package should appear only once."
        )
```

File: tests/integration/test_github_workflows.py [hunk 9]
```diff
@@ -418,31 +409,16 @@ def test_pr_agent_no_duplicate_setup_steps(self, pr_agent_workflow: Dict[str, An
         )
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
        )
```

File: tests/integration/test_github_workflows.py [hunk 10]
```diff
@@ -418,32 +415,39 @@ def test_pr_agent_no_duplicate_setup_steps(self, pr_agent_workflow: Dict[str, An
...
+                f"fetch-depth should be an integer, got {type(fetch_depth).__name__}"
+            )
+            # Reject negative integers
+            assert fetch_depth >= 0, "fetch-depth cannot be negative"
+            step_with = step.get("with", {})
+            # It's acceptable for fetch-depth to be omitted entirely
+            if "fetch-depth" not in step_with:
+                continue
+            fetch_depth = step_with["fetch-depth"]
+            # Reject non-integer types (including strings)
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
            assert fetch_depth >= 0, "fetch-depth cannot be negative"
            step_with = step.get("with", {})
            # It's acceptable for fetch-depth to be omitted entirely
            if "fetch-depth" not in step_with:
                continue
            fetch_depth = step_with["fetch-depth"]
            # Reject non-integer types (including strings)
```

File: tests/integration/test_github_workflows.py [hunk 11]
```diff
@@ -339,27 +341,22 @@ def test_pr_agent_checkout_has_token(self, pr_agent_workflow: Dict[str, Any]):
...
-    Parameters:
-        pr_agent_workflow (Dict[str, Any]): Parsed YAML mapping for the pr-agent workflow; expected to contain a "jobs" mapping with a "pr-agent-trigger" job.
-    """
-    trigger_job = pr_agent_workflow["jobs"]["pr-agent-trigger"]
-            pr_agent_workflow (Dict[str, Any]): Parsed YAML mapping for the pr-agent workflow; expected to contain a "jobs" mapping with a "review" job.
-        """
+        """Asserts the workflow's trigger job includes a setup-python step."""
+
         review_job = pr_agent_workflow["jobs"]["pr-agent-trigger"]
         steps = review_job.get("steps", [])
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
        review_job = pr_agent_workflow["jobs"]["pr-agent-trigger"]
        steps = review_job.get("steps", [])
```

File: tests/integration/test_github_workflows.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
        for key in env_dict.keys():
            # Ensure all characters are either alphanumeric or underscore
            is_valid_chars = all(c.isalnum() or c == '_' for c in key)
            if not key.isupper() or not is_valid_chars:
                invalid.append(key)
        return invalid
```

File: tests/integration/test_github_workflows.py [hunk 12]
```diff
@@ -418,31 +409,16 @@ def test_pr_agent_no_duplicate_setup_steps(self, pr_agent_workflow: Dict[str, An
...
-        
+
         checkout_steps = [
-            s for s in steps 
+            s for s in steps
             if s.get("uses", "").startswith("actions/checkout")
         ]
-        
+
         for step in checkout_steps:
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
        for step in checkout_steps:
```

File: tests/integration/test_github_workflows.py [hunk 13]
```diff
@@ -342,24 +344,13 @@ def test_pr_agent_checkout_has_token(self, pr_agent_workflow: Dict[str, Any]):
             assert "token" in step_with, "Checkout step should specify a token"
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
            assert "token" in step_with, "Checkout step should specify a token"
```

File: tests/integration/test_requirements_dev.py [hunk 14]
```diff
@@ -274,32 +275,213 @@
...
+        """Test that PyYAML and types-PyYAML are on separate lines (not combined)."""
+        assert REQUIREMENTS_FILE.exists(), "requirements-dev.txt not found"
+        
+        with open(REQUIREMENTS_FILE, 'r') as f:
+            lines = [line.strip() for line in f if line.strip() and not line.startswith('#')]
+        
+        # Find lines that start with either PyYAML or types-PyYAML
+        pyyaml_lines = [line for line in lines if line.startswith('PyYAML') or line.startswith('types-PyYAML')]
+        
+        assert len(pyyaml_lines) == 1, (
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_dev.py]:
```suggestion
        assert len(pyyaml_lines) == 1, (
```

File: tests/integration/test_requirements_dev.py [hunk 15]
```diff
@@ -160,14 +159,16 @@
...
-    def test_version_format_valid(self, requirements: List[Tuple[str, str]]):
-        """Test that version specifications use valid format."""
-    from packaging.specifiers import SpecifierSet
     def test_version_format_valid(self, requirements: List[Tuple[str, str]]):
         """Test that version specifications use valid PEP 440 format."""
         for pkg, ver_spec in requirements:
+            if ver_spec:
+                try:
+                    SpecifierSet(ver_spec)
+                except InvalidSpecifier as e:
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_dev.py]:
```suggestion
                except InvalidSpecifier as e:
```

File: tests/integration/test_requirements_dev.py [hunk 16]
```diff
@@ -274,32 +275,213 @@
...
+        with open(REQUIREMENTS_FILE, 'r') as f:
+            lines = [line.strip() for line in f if line.strip() and not line.startswith('#')]
+        
+        # Find lines that start with either PyYAML or types-PyYAML
+        pyyaml_lines = [line for line in lines if line.startswith('PyYAML') or line.startswith('types-PyYAML')]
+        
+        assert len(pyyaml_lines) == 1, (
+            f"Should have exactly 1 line for PyYAML, found {len(pyyaml_lines)}"
+        )
+        assert len(types_pyyaml_lines) == 1, (
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_requirements_dev.py]:
```suggestion
        assert len(types_pyyaml_lines) == 1, (
```

**Review by mohavro — COMMENTED** [review]
merge

**Review by mohavro — COMMENTED** [review]
Ready for merge

**Review by mohavro — COMMENTED** [review]
Repeat all PR reviews

File: tests/integration/test_github_workflows.py [hunk 17]
```diff
@@ -376,12 +399,16 @@ def test_pr_agent_has_node_setup(self, pr_agent_workflow: Dict[str, Any]):
         assert len(node_steps) > 0, "Review job must set up Node.js"
     
     def test_pr_agent_python_version(self, pr_agent_workflow: Dict[str, Any]):
+"""
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_github_workflows.py]:
```suggestion
"""
```

File: tests/integration/test_workflow_requirements_integration.py [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+                
+                # Look for pip install or requirements file installation
+                if ('pip install' in run_cmd.lower() or 
+                    'requirements' in run_cmd.lower() or
+                    'requirements' in step_name):
+                    has_install_step = True
+                    break
+            
+            if has_install_step:
+                break
```
[the remark this answers is not in the document]
Comment (mohavro) [reply tests/integration/test_workflow_requirements_integration.py]:
```suggestion
        for job_name, job in workflow.get('jobs', {}).items():
            for step in job.get('steps', []):
                run_cmd = step.get('run', '')
                step_name = step.get('name', '').lower()
                
                # Look for pip install or requirements file installation
                if ('pip install' in run_cmd.lower() or 
                    'requirements' in run_cmd.lower() or
                    'requirements' in step_name):
                    has_install_step = True
                    break
            
            if has_install_step:
                break
```

**Review by mohavro — COMMENTED** [review]
Review Summary
- Aligns tests with 'pr-agent-trigger' job and uses yaml import skipping appropriately
- Recommend additional changes to restore security coverage, ensure cross-platform reliability, and reduce duplication

Recommendations
1) Security scanning: add back a dedicated test to detect hardcoded secrets/tokens in `with`, `env`, and `run` content; only allow values via `${{ secrets... }}` or empty
2) Windows UTF-8 reads: specify `encoding='utf-8'` when reading workflow/documentation files to avoid `UnicodeDecodeError` on Windows
3) Dependencies: `tests/integration/test_workflow_requirements_integration.py` uses `packaging`; add `packaging>=24.0` to `requirements-dev.txt` or guard with `pytest.importorskip('packaging')`
4) Helper duplication: factor `check_env_vars` into a single helper and reuse
5) Runner and action pinning: assert pinned runner (e.g., `ubuntu-latest` or specific LTS) and keep assertions that third-party actions are pinned to a version
6) fetch-depth robustness: add tests that reject negative integers and non-integer types; allow omission
7) Message consistency: update messages and test names to reference `'pr-agent-trigger'` consistently (not `'Review job'`)
8) Workflows dir resolution: use a project-root discovery helper to derive `.github/workflows` so tests are robust to cwd changes

CI Status
- Current checks show failures in python-lint, Vercel deploy, and frontend build; after addressing the above, please re-run CI. I can review the failures separately if desired.

**Comment by mohavro** [comment]
## Syntax Errors Found in `tests/integration/test_github_workflows.py`

The PR has syntax errors that prevent the Python file from compiling:

### Issues Identified:

1. **Line 448**: Malformed docstring with duplicate content and incorrect triple-quote placement
2. **Multiple lines**: Apostrophes in docstrings causing unterminated string literals (e.g., "job's", "don't", "workflow's")
3. **Duplicate code blocks**: Lines 1135-1165 and 1194-1224 contain duplicate `check_env_vars` function definitions
4. **Line 2901**: Unterminated triple-quoted string literal

### Recommended Fixes:

1. Remove apostrophes from all docstrings (replace with "do not", "does not", etc.)
2. Fix the malformed docstring at line 448
3. Remove duplicate code blocks
4. Ensure all triple-quoted strings are properly closed

I will proceed to fix these issues and push the corrections to the branch.

**Comment by mohavro** [comment]
## Summary of Review

After extensive analysis, PR #239 has multiple **critical syntax errors** that prevent the Python files from compiling:

### Critical Issues:

1. **Malformed docstring at line 448** - Incorrectly formatted triple quotes with duplicate content
2. **Unclosed string literals** - Multiple apostrophes in docstrings throughout the file ("don't", "job's", "workflow's", etc.)
3. **Duplicate code blocks** - Functions `check_env_vars` defined multiple times (lines 1135-1165 and 1194-1224)
4. **File truncation** - The file appears to end abruptly at line 2919 with unclosed strings

### Recommendation:

**Do not merge this PR in its current state.** The changes need to be completely reworked. 

### Suggested Approach:

1. Start with a clean version from the base branch (origin/main)
2. Apply only the essential PyYAML import changes:
   - Replace `import yaml` with `yaml = pytest.importorskip("yaml")`
3. Avoid making extensive changes to docstrings and comments
4. Thoroughly test syntax with `python -m py_compile` before committing

### Alternative:

Close this PR and create a new, focused PR that:
- Makes minimal changes
- Only addresses the PyYAML import issue
- Passes syntax validation
- Can be reviewed and merged quickly

The current PR has accumulated too many issues to fix incrementally.

**Review by mohavro — COMMENTED** [review]
a few more changes left in review comments then ready for merge
